src/NuGet.Protocol.Catalog/Models/BasePackageDependencyGroup.cs

```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Newtonsoft.Json;
+
+namespace NuGet.Protocol.Catalog
+{
+    public class BasePackageDependencyGroup<TDependency> where TDependency : PackageDependency
+    {
+        [JsonProperty("@id")]
```

Comment:
Another thought--all entities in SPARQL/JSON-LD/whatever have `@id` and `@type` properties. Should we make a shared interface or base class that all of these types extend, so that these fields don't have to be declared repeatedly?

Reply:
I believe inheritance for the sole purpose of DRY is a bad pattern. I've seen inheritance used WAY too often as a method of code sharing where composition really should have been used. In my experience, I've found inheritance should be used a lot more sparingly than it is in practice.

In this PR, I only added inheritance in cases where there is benefit moving forward and the super-sub relationship is strong in the _ideal_ mental model for our entities. For the package dependency group in catalog vs. registration, this is clear subset/superset relationship. I don't even feel strongly about this one. I wouldn't feel too bad about cutting the relationship and introducing mapping code. For alternate package and package dependency, I don't feel as strongly but I think it's a bad idea to add inheritance just as a DRY convenience. What scenario would we be enabling by clubbing these together?

For these JSON models I am not convinced that inheritance actually makes our lives better. I think it's just as likely that we have unintended properties leaking to places we don't want them.

For `@id` and `@type` I don't think it's right to add a base class just so we can avoid defining these unfortunate properties multiple times. For most of the `@id` and `@type` cases, they are there as an artifact of our JSON-LD implementation not because we actually think they are useful or provide crucial new information. Consider the `@id` of a package dependency group or the `@type` of deprecations.

On a related note, I only added the `ICommitted` interface because it enables a single method to set the commit ID and commit timestamp for N different classes. I think this is a legitimate use of an interface but would be an inappropriate use for a base class.

If you want to throw all of my anti-OOP mumbo-jumbo out the window my point can be simplified to this: I wanted the most granular control possible for us to match the existing JSON shape in a very predictable way. Inheritance makes this harder.

Reply:
I wouldn't. `BaGet.Protocol` renames the `@id` property to more logical names like `CatalogLeafUrl` to describe the destination object. We can't do this if the property names are restricted by an interface

Reply:
I don't think it's necessarily just a "DRY convenience"--they are both links to specific versions/ranges of a package. I can imagine some code that finds a package in registration from a `PackageDependency`, which could be reused for both finding dependency metadata and for finding alternate package metadata. The DRY aspect is just a bonus. Worst case scenario, if the types diverge down the line, we can split the types apart, which shouldn't be a very difficult change to make. I can agree that adding a base class is overdesigning, but the V3 docs have literally defined an alternate package as "The package dependency that should be used instead".

I would argue, regarding `@id` and `@type`, that the fact that they are "unfortunate" and generally unused is precisely the reason why we should put them on a single base class or interface. When we finally decide to fully get rid of them, we can remove their declaration from a single place and avoid updating N files. Even if we do have to update each file, for some other reason, it will be very clear to us which files will need to be updated.

I agree that too much OOP can slow development, but I think for these two changes we have a good reason to do the inheritance. Ultimately, however, I'm fine with this PR moving forward without them.

Reply:
I didn't remember that an alternate package is a package dependency. I would have pushed back on that at the protocol stage if I caught it. I think from a domain perspective they are different things and adding properties to one won't always mean we want that property in the other. One critical difference is that a range can be `*` in alternate package but package dependencies cannot have `*`. They are different models IMO.